### PR TITLE
Simulator

### PR DESCRIPTION
**Pull request by ptesavol**
This is the first version of streamr-simulator that replaces the connection layer with a simulated one with the help of Jest moduleNameMapper, and funnels all network traffic through the Simulator singleton. You can now run the network tests  in simulator using the command "npm run test-simulator" (only a few tests, such as browser tests have been excluded). 

The current version of the simulator establishes connections and passes messages without any delay (in the same tick). This extreme (and unrealistic) setting was used deliberately in order to locate and remove the risky "first do something and only then start waiting for the results" pattern from the tests. The Simulator class itself can easily be replaced using the moduleNameMapper with other implementations tailored for other use cases.  

Next versions of the Simulator will add an API to the Simulator class with features such as fetching the current topology, and adding hooks to the send and connect functions.

### CODE REVIEW

File: packages/network/src/connection/WebRtcConnection.ts [hunk 1]
```diff
@@ -275,6 +277,7 @@ export abstract class WebRtcConnection extends ConnectionEmitter {
             if (this.pingAttempts >= this.maxPingPongAttempts) {
                 this.baseLogger.debug(`failed to receive any pong after ${this.maxPingPongAttempts} ping attempts, closing connection`)
                 this.close(new Error('pong not received'))
```
Comment (timoxley) [inline packages/network/src/connection/WebRtcConnection.ts:283]: Should this `clearTimeout(this.pingTimeoutRef)` perhaps?

File: packages/network/src/connection/WebRtcEndpoint.ts [hunk 2]
```diff
@@ -378,11 +378,16 @@ export class WebRtcEndpoint extends EventEmitter implements IWebRtcEndpoint {
             this.rtcSignaller.sendRtcConnect(routerId, connection.getPeerId())
         }
 
-        const deferredAttempt = connection.getDeferredConnectionAttempt()
+        const deferredAttempt = connection.getDeferredConnectionAttempt() 
+        
+        if (connection.getLastState() == 'connected') {
+            return Promise.resolve(targetPeerId)
```
Comment (timoxley) [inline packages/network/src/connection/WebRtcEndpoint.ts]: Perhaps eslint/typescript is complaining but you shouldn't need to wrap returns in `Promise.resolve`. 
`return targetPeerId` will do the same thing as `return Promise.resolve(targetPeerId)` inside an `async` function.

File: packages/network/src/connection/WebRtcEndpoint.ts [hunk 3]
```diff
@@ -378,11 +378,16 @@ export class WebRtcEndpoint extends EventEmitter implements IWebRtcEndpoint {
...
+        }
         if (deferredAttempt) {
             return deferredAttempt.getPromise()
-        } else {
-            throw new Error(`disconnected ${connection.getPeerId()}`)
+        } 
+        else { 
+            return Promise.reject(new WebRtcError(`disconnected ${connection.getPeerId()}`))
```
Comment (timoxley) [inline packages/network/src/connection/WebRtcEndpoint.ts]: Same here, just `throw`, it'll be wrapped in a rejection.

File: package.json [hunk 4]
```diff
@@ -28,6 +28,7 @@
         "@commitlint/cli": "^12.1.4",
         "@commitlint/config-conventional": "^12.1.4",
         "@manypkg/cli": "^0.19.1",
+        "ci": "^2.1.1",
```
Comment (timoxley) [inline package.json]: ?

File: packages/network/package.json [hunk 5]
```diff
@@ -54,9 +55,10 @@
     "geoip-lite": "^1.4.2",
     "heap": "^0.2.6",
     "ipaddr.js": "^2.0.1",
+    "leaked-handles": "^5.2.0",
```
Comment (juslesan) [inline packages/network/package.json]: clean leaked-handles debugging

File: packages/network/test/integration/webrtc-error-scenarios-during-signalling.test.ts [hunk 6]
```diff
@@ -1,11 +1,27 @@
-import { runAndWaitForEvents, runAndRaceEvents, waitForEvent } from 'streamr-test-utils'
+/*
+require('leaked-handles').set({
```
Comment (juslesan) [inline packages/network/test/integration/webrtc-error-scenarios-during-signalling.test.ts]: clean leaked-handles debugging

File: packages/network/test/integration/webrtc-error-scenarios-during-signalling.test.ts [hunk 7]
```diff
@@ -1,11 +1,27 @@
...
+
+import { runAndWaitForEvents } from 'streamr-test-utils'
 
 import { Tracker } from '../../src/logic/tracker/Tracker'
 import { NetworkNode } from '../../src/logic/node/NetworkNode'
 import { createNetworkNode, startTracker } from '../../src/composition'
 import { Event as NodeEvent } from '../../src/logic/node/Node'
 import { Event as NodeToTrackerEvent } from '../../src/protocol/NodeToTracker'
 
+/*
```
Comment (juslesan) [inline packages/network/test/integration/webrtc-error-scenarios-during-signalling.test.ts]: clean leaked-handles debugging

File: packages/network/test/simulator/ServerWsEndpoint_simulator.test.ts [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+import { runAndWaitForEvents } from 'streamr-test-utils'
+import { DisconnectionCode, DisconnectionReason, Event } from '../../src/connection/ws/AbstractWsEndpoint'
+
+const wssPort1 = 7777
+
+describe('ServerWsEndpoint', () => {
+    let serverWsEndpoint: ServerWsEndpoint | undefined = undefined
+    let clientWsEndpoint: NodeClientWsEndpoint | undefined = undefined
+
+    /*
```
Comment (juslesan) [inline packages/network/test/simulator/ServerWsEndpoint_simulator.test.ts]: Clean up commented blocks in this test file

File: packages/network/src/simulator/ServerWsEndpoint_simulator.ts [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+                        this.logger.error('error closing http server: %s', err)
+                        reject(err)
+                    } else {
+                        resolve()
+                    }
+                })
+            })
+        }
+        /*
+        return new Promise((resolve, reject) => {
```
Comment (juslesan) [inline packages/network/src/simulator/ServerWsEndpoint_simulator.ts:237]: clean up commented blocks of code

File: packages/network/src/simulator/ServerWsEndpoint_simulator.ts [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+                this.onReceive(connection, data.toString())
+            }
+        }
+        else {
+            const connection = this.getConnectionByPeerId(fromInfo.peerId) as AbstractWsConnection
+            this.onReceive(connection, data)
+        }
+    }
+
+    /****************** Called by Simulator ends *******/
```
Comment (juslesan) [inline packages/network/src/simulator/ServerWsEndpoint_simulator.ts:148]: clean up commented blocks of code

**Review by juslesan — CHANGES_REQUESTED** [review]
Good stuff in general!

There's a lot of comment blocks consisting of unused code that should be cleaned up before merging.

**Pull request merged**